daemon/state.rs

```diff
@@ -0,0 +1,99 @@
+use atomicwrites::{AtomicFile, OverwriteBehavior};
+use serde::{Deserialize, Serialize};
+use std::io::Write;
+use std::path::Path;
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct State {
+    pub total_jobs_completed: u64,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self {
+            total_jobs_completed: 0,
+        }
+    }
+}
+
+pub fn load(state_file: &Path) -> State {
+    log::debug!("Loading initial state from {}", state_file.display());
+    match std::fs::read_to_string(state_file) {
+        Err(err) => {
+            if err.kind() != std::io::ErrorKind::NotFound {
+                log::warn!(
+                    "Cannot read initial state from {}: {}",
+                    state_file.display(),
+                    err
+                );
+            }
+            return State::default();
+        }
+        Ok(data) => match serde_json::from_str::<State>(&data) {
+            Err(err) => {
+                log::warn!(
+                    "Cannot parse initial state from {}: {}",
+                    state_file.display(),
+                    err
+                );
+                return State::default();
+            }
+            Ok(state) => {
+                log::debug!("Initial state: {:?}", state);
+                return state;
+            }
+        },
+    }
+}
+
+pub fn store(state_file: &Path, state: &State) {
+    if let Some(parent) = state_file.parent() {
```

Comment:
checking for existing parent dir on every write is expensive, what do you think about we instead let the write fail, and only then create the parent dir?

Reply:
Great suggestion! See 7e7d49a